File: ruby/perfect-numbers/perfect_numbers.rb
module PerfectNumber
  def self.classify(n)
    raise RuntimeError if n < 0

    case aliquot_sum(n) <=> n
    when 0 then "perfect"
    when 1 then "abundant"
    when -1 then "deficient"
    end
  end

  def self.aliquot_sum(n)
    (1..n/2)
      .select { |i| n % i == 0 }
      .sum
  end
end

File: javascript/perfect-numbers/perfect-numbers.js
export const classify = (n) => {
    if (n <= 0)
        throw "Classification is only possible for natural numbers.";

    let aliquot_sum = 0;
    let upper = parseInt(n / 2);

    for (let i = 1; i <= upper; ++i)
        if (n % i == 0)
            aliquot_sum += i;

    if (aliquot_sum > n)
        return "abundant";
    else if (aliquot_sum < n)
        return "deficient";
    else
        return "perfect";
};
